api/proto/auth/client.go

```diff
@@ -0,0 +1,186 @@
+package auth
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"sync"
+	"sync/atomic"
+
+	"github.com/gravitational/roundtrip"
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+)
+
+var log = logrus.WithFields(logrus.Fields{
+	trace.Component: teleport.ComponentAuth,
+})
+
+// Client is HTTP Auth API client. It works by connecting to auth servers
+// via HTTP.
+//
+// When Teleport servers connect to auth API, they usually establish an SSH
+// tunnel first, and then do HTTP-over-SSH. This client is wrapped by TunClient
+// in lib/auth/tun.go
+type Client struct {
+	sync.Mutex
+	ClientConfig
+	grpc AuthServiceClient
+	conn *grpc.ClientConn
+	// closedFlag is set to indicate that the services are closed
+	closedFlag int32
+}
+
+// NewClient establishes a gRPC connection to an auth server.
+func NewClient() (*Client, error) {
+	tlsConfig, err := PathCreds("certs/api-admin")
+	if err != nil {
+		return nil, fmt.Errorf("Failed to setup TLS config: %v", err)
+	}
+
+	// replace 127.0.0.1:3025 (default) with your auth server address
+	addrs := []utils.NetAddr{{Addr: "127.0.0.1:3025"}}
+	clientConfig := ClientConfig{Addrs: addrs, TLS: tlsConfig}
+
+	return NewTLSClient(clientConfig)
+}
+
+// NewTLSClient returns a new TLS client that uses mutual TLS authentication
+// and dials the remote server using dialer. Connection is loaded lazily.
+func NewTLSClient(cfg ClientConfig, params ...roundtrip.ClientParam) (*Client, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &Client{ClientConfig: cfg}, nil
+}
+
+// PathCreds loads mounted creds from path, detects reloads and updates the grpc transport
+func PathCreds(path string) (*tls.Config, error) {
```

Comment:
Is this new functionality? Let's keep this PR narrow and focused, just moving code around. We can add the credential loading logic in another PR.

Reply:
No this is essentially what I had in examples/goclient/client.go, which is being replaced by this client.